### PR TITLE
Implement fragment lighting in the sw renderer (take 2)

### PR DESCRIPTION
**Pull request by wwylele**
Original PR #2766. Addressed comments that were left there. Didn't add new features (that were added to the gl rasterizer recently).

One of the comments was not addressed:
> yuriks: This entire function is in dire need of some splitting up in separate functions (that calculate the influence for one light, calculate sub-parts of the equation, etc.). It is extremely hard to follow what values are going where right now.

I hesitated doing this, because each part of equation interlocks to each other very much, and it will end up with functions with long parameter lists, such as `GetLutValue(lighting, lighting_state, light_vector, norm, view, input, abs, scale_enum, sampler)`. Making them local lambda is effectively the same as is and doesn't do much either.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/2822)

### CODE REVIEW

**Comment by B3n30** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by yuriks** [comment]
:lgtm:

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 118 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqR21kC6M7lqrFfcrXb:-KqR21kC6M7lqrFfcrXc:b1pnbkh) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L118)):*
> ```cpp
> }
> 
> static float LookupLightingLut(const Pica::State::Lighting& lighting, size_t lut_index, u8 index,
> ```

Any reason you didn't move lighting code to a separate file? Still worth doing even if you don't split the function imo.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 144 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqR2R188hDgythu5OG3:-KqR2R188hDgythu5OG4:b-c5f2ko) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L144)):*
> .Normalized()
Should probably just normalize it after it's interpolated, in the caller function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822#-:-KqR2fjOp7A9cGigNxP1:bnfp4nl)*

**Comment by vonchenplus** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/common/quaternion.h, line 35 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-Kq6gAMi9rqH9IRYZev-:-Kq6gAMi9rqH9IRYZev0:b-1dahtg) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/common/quaternion.h#L35)):*
> ```objc
> 
>     Quaternion<T> Normalized() const {
>         T length = std::sqrt(xyz.Length2() + w * w);
> ```

divide or mod by T::zero

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by vonchenplus** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 144 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqSFq3foSaK3ITBFkDL:-KqSFq3foSaK3ITBFkDM:b-18rk5g) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L144)):*
> ```cpp
> 
>     // Use the normalized the quaternion when performing the rotation
>     auto normal = Math::QuaternionRotate(normquat.Normalized(), surface_normal);
> ```

need Normalized

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by vonchenplus** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 121 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqSHQ40sXObPk6MRfSp:-KqSHQ40sXObPk6MRfSq:bsa8f8j) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L121)):*
> ```cpp
>                                float delta) {
>     ASSERT_MSG(lut_index < lighting.luts.size(), "Out of range lut");
>     ASSERT_MSG(index < lighting.luts[0].size(), "Out of range index");
> ```

lighting.luts[lut_index].size instead of lighting.luts[0]

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by vonchenplus** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 209 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqSI0JdeBEFaP4qhBXR:-KqSI0JdeBEFaP4qhBXS:bjboauc) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L209)):*
> ```cpp
>                 LOG_CRITICAL(HW_GPU, "Unknown lighting LUT input %u\n", static_cast<u32>(input));
>                 UNIMPLEMENTED();
>                 result = 0.0f;
> ```

return value immediately

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by wwylele** [comment]
Review status: 4 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/common/quaternion.h, line 35 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-Kq6gAMi9rqH9IRYZev-:-KqZ5zlFxWnWlo2KiM3J:byzwkac) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/common/quaternion.h#L35)):*
<details><summary><i>Previously, vonchenplus (Feng Chen) wrote…</i></summary><blockquote>

divide or mod by T::zero
</blockquote></details>

This isn't intended to be used for integral type `T`, and for floating type it has defined behaviour. The caller should mind for checking if length is zero - if it doesn't check, like in this case, it means it assumes it will never be zero, or when it is zero there is no intended behaviour (as 3DS will also glitch out in this case). And anyway, this keeps the same style as `Normalized` function in `vector_math`

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 118 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqR21kC6M7lqrFfcrXb:-KqZOCN2QRclPdjEMeq5:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L118)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

Any reason you didn't move lighting code to a separate file? Still worth doing even if you don't split the function imo.
</blockquote></details>

Done.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 121 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqSHQ40sXObPk6MRfSp:-KqZOCnxdIfKNLKSdrUt:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L121)):*
<details><summary><i>Previously, vonchenplus (Feng Chen) wrote…</i></summary><blockquote>

lighting.luts[lut_index].size instead of lighting.luts[0]
</blockquote></details>

Done.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 144 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqR2R188hDgythu5OG3:-KqZ9P1mS1u4bZO1jJym:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L144)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

> .Normalized()
Should probably just normalize it after it's interpolated, in the caller function.
</blockquote></details>

Done.

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 144 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqSFq3foSaK3ITBFkDL:-KqZ9PXRGe7r9lSbj8Gq:b-57hnri) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L144)):*
<details><summary><i>Previously, vonchenplus (Feng Chen) wrote…</i></summary><blockquote>

need Normalized
</blockquote></details>

What do you mean exactly? If you mean `normal` needs `Normalized`, `QuaternionRotate` doesn't change the length of the vector, and as `surface_normal` has a length of one, `normal` will have the same unit length

---

*[src/video_core/swrasterizer/rasterizer.cpp, line 209 at r1](https://reviewable.io:443/reviews/citra-emu/citra/2822#-KqSI0JdeBEFaP4qhBXR:-KqZJQnYGSZvCZeRp49n:bg08t2m) ([raw file](https://github.com/citra-emu/citra/blob/4feff63ffaec4d62d5bdfc85968cc99298907767/src/video_core/swrasterizer/rasterizer.cpp#L209)):*
<details><summary><i>Previously, vonchenplus (Feng Chen) wrote…</i></summary><blockquote>

return value immediately
</blockquote></details>

IMO it will be more confusing due to inconsistency with other `case`s, and after all `LightingLutInput` are fully implemented, this `default` is unlikely to execute. The PR is not written by me originally, so I don't want to make unnecessary changes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by yuriks** [comment]
Reviewed 1 of 1 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/video_core/CMakeLists.txt, line 58 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2822#-Kqa0ydomU-oVFvtDaWK:-Kqa0ydomU-oVFvtDaWL:bxixpq1) ([raw file](https://github.com/citra-emu/citra/blob/eda28266fb1f0eb96a2096cadb41b62db3dc2d2e/src/video_core/CMakeLists.txt#L58)):*
> ```CMake
>             shader/shader_interpreter.h
>             swrasterizer/clipper.h
>             swrasterizer/fragment_lighting.h
> ```

nit: This can be named just `lighting`, since it's implicit that any lighting done by the rasterizer is fragment lighting. Keeps the names nice and short. :D

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by wwylele** [comment]
Review status: 4 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/video_core/CMakeLists.txt, line 58 at r4](https://reviewable.io:443/reviews/citra-emu/citra/2822#-Kqa0ydomU-oVFvtDaWK:-KqbK8ROg7hhqRnq5stS:b-896fix) ([raw file](https://github.com/citra-emu/citra/blob/eda28266fb1f0eb96a2096cadb41b62db3dc2d2e/src/video_core/CMakeLists.txt#L58)):*
<details><summary><i>Previously, yuriks (Yuri Kunde Schlesner) wrote…</i></summary><blockquote>

nit: This can be named just `lighting`, since it's implicit that any lighting done by the rasterizer is fragment lighting. Keeps the names nice and short. :D
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by B3n30** [comment]
Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822)*

**Comment by B3n30** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/citra-emu/citra/2822#-:-KqqwKOK7CAQLTfHSce8:bnfp4nl)*

**Comment by wwylele** [comment]
> \<wwylele> yuriks: 2822 and 2856 sign off pleeease?
> \<yuriks\> I asked you if 2822 was ready for merging the other day... Feel free to merge
